compiler/rustc_const_eval/src/const_eval/mod.rs

```diff
@@ -90,19 +106,91 @@ fn const_to_valtree_inner<'tcx>(
         // Technically we could allow function pointers (represented as `ty::Instance`), but this is not guaranteed to
         // agree with runtime equality tests.
         ty::FnPtr(_) | ty::RawPtr(_) => None,
-        ty::Ref(..) => unimplemented!("need to use deref_const"),
+
+        ty::Ref(_, ref_ty, _) if place.layout.ty.is_slice() => {
+            match ecx.try_read_immediate_from_mplace(&place) {
+                Ok(Some(imm)) => {
+                    // `imm` is a ScalarPair. We try to get the underlying bytes behind that
+                    // fat pointer for string slices and slices of integer types. For any other
+                    // slice types we use `branches` to recursively construct the Valtree.
+
+                    if use_bytes_for_ref(*ref_ty) {
+                        let (alloc, range) = ecx.get_alloc_from_imm_scalar_pair(imm);
+                        let alloc_bytes = match alloc.get_bytes(&ecx.tcx, range) {
+                            Ok(bytes) => bytes,
+                            Err(_e) => return None,
+                        };
+                        debug!(?alloc_bytes);
```

Comment:
Shouldn't these allocations already be interned? So you should be able to get `alloc_bytes` as `'tcx` lifetime here.

Reply:
But I guess this will all go away anyway if we don't have `ValTree::SliceOrStr`. I think we might want to have that eventually, but it should definitely be a separate PR from adding support for references in the first place -- reviewing both of these changes together is a nightmare.

Reply:
[`get_bytes`](https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_middle/mir/interpret/allocation.rs.html#259-261) doesn't return a slice with a `'tcx` lifetime annotation.

Reply:
`get_bytes` uses the lifetime of the `Allocation`. If you work on interned data, you should be able tog et your hands on a `&'tcx Allocation`.

But as you say that's for a separate PR, as this PR shouldn't do `SliceOrStr`.